### PR TITLE
1.16 nether, new blocks and new structures

### PR DESCRIPTION
**Pull request by BlusteryS**
This pull adds vanilla nether and some structures.

### CODE REVIEW

**Comment by lt-name** [comment]
In nk projects, it is common to use 4 spaces instead of a tab for indentation
I think it's better to change to 4 spaces, keep the same as before

**Comment by BlusteryS** [comment]
In many files (not only mine) there are tabs, replace them with spaces?

**Comment by BlusteryS** [comment]
image

**Comment by lt-name** [comment]
Try to ensure that the new content uses four spaces

**Comment by lt-name** [comment]
The default world generator doesn't seem to work, only the void

**Comment by BlusteryS** [comment]
> The default world generator doesn't seem to work, only the void

image
image

File: src/main/java/cn/nukkit/level/Level.java [hunk 1]
```diff
@@ -39,29 +51,80 @@
...
+import cn.nukkit.math.SimpleAxisAlignedBB;
+import cn.nukkit.math.Vector2;
+import cn.nukkit.math.Vector3;
+import cn.nukkit.math.Vector3f;
 import cn.nukkit.metadata.BlockMetadataStore;
 import cn.nukkit.metadata.MetadataValue;
 import cn.nukkit.metadata.Metadatable;
 import cn.nukkit.nbt.NBTIO;
-import cn.nukkit.nbt.tag.*;
-import cn.nukkit.network.protocol.*;
```
Comment (lt-name) [inline src/main/java/cn/nukkit/level/Level.java:56]: Like this, I think you can continue to use * instead of citing item by item

**Comment by lt-name** [comment]
> > The default world generator doesn't seem to work, only the void
> 
> image image

I tried again, it's the plugin's influence, sorry

File: src/main/java/cn/nukkit/level/Level.java [hunk 2]
```diff
@@ -2969,36 +2970,31 @@ public void generateChunkCallback(int x, int z, BaseFullChunk chunk) {
...
-                    loader.onChunkPopulated(chunk);
-                }
-            }
-        } else if (this.chunkGenerationQueue.containsKey(index) || this.chunkPopulationLock.containsKey(index)) {
-            this.chunkGenerationQueue.remove(index);
-            this.chunkPopulationLock.remove(index);
+            setChunk(x, z, chunk, false);
+            chunk = getChunk(x, z, false);
+            if (chunk != null && (oldChunk == null || !isPopulated) && chunk.isPopulated() && chunk.getProvider() != null) {
+                new ChunkPopulationTask(this, chunk, getDimension() == Level.DIMENSION_NETHER ? Generator.netherPopulator : Generator.overworldPopulators).start();
```
Comment (lt-name) [inline src/main/java/cn/nukkit/level/Level.java:2988]: Something doesn't seem right here.

Comment (BlusteryS) [reply src/main/java/cn/nukkit/level/Level.java]: resolved

**Review by lt-name — APPROVED** [review]

**Pull request merged**